### PR TITLE
plugin.video.themoviedb.helper 4.10.14

### PR DESCRIPTION
**Pull request by jurialmunkey**
### Description
Significant code refactor to allow for lazy module importing, improved caching, and faster item building.

### Checklist:

- [x] My code follows the [add-on rules](http://kodi.wiki/view/Add-on_rules) and [piracy stance](http://kodi.wiki/view/Official:Forum_rules#Piracy_Policy) of this project. 
- [x] I have read the [CONTRIBUTING](https://github.com/xbmc/repo-plugins/blob/master/CONTRIBUTING.md) document
- [x] Each add-on submission should be a single commit with using the following style: [plugin.video.foo] v1.0.0

Additional information :
- Submitting your add-on to this specific branch makes it available to any Kodi version equal or higher than the branch name with the applicable Kodi dependencies limits.
- [add-on development](http://kodi.wiki/view/Add-on_development) wiki page.
- Kodi [pydocs](http://kodi.wiki/view/PyDocs) provide information about the Python API
- [PEP8](https://www.python.org/dev/peps/pep-0008/) codingstyle which is considered best practice but not mandatory.
- This add-on repository has automated code guideline check which could help you improve your coding. You can find the results of these check at [Codacy](https://www.codacy.com/app/Kodi/repo-plugins/dashboard). You can create your own account as well to continuously monitor your python coding before submitting to repo.
- Development questions can be asked in the [add-on development](http://forum.kodi.tv/forumdisplay.php?fid=26) section on the Kodi forum.
- If you see no activity on your PR after a week (so at least one weekend has passed) then please go to the #kodi-dev freenode IRC channel to reach out to the team

### CODE REVIEW

**Comment by jurialmunkey** [comment]
Force pushing to bring in line with latest master as PR is now significantly out of date again.

**Comment by jurialmunkey** [comment]
Force pushing once again to bring up-to-date with latest. Just ping me when someone actually does get around to reviewing this...

File: plugin.video.themoviedb.helper/addon.xml [hunk 1]
```diff
@@ -15,43 +16,47 @@ provider-name="jurialmunkey">
...
     </item>
-    <item library="context-sorting.py">
+    <item library="context.py" args="sorting">
       <label>$ADDON[plugin.video.themoviedb.helper 32309]</label>
       <visible>!String.IsEmpty(ListItem.Property(tmdbhelper.context.sorting))</visible>
     </item>
   </menu>
 </extension>
 <extension point="xbmc.addon.metadata">
+  <reuselanguageinvoker>false</reuselanguageinvoker>
```
Comment (enen92) [inline plugin.video.themoviedb.helper/addon.xml:50]: not needed but ok

File: plugin.video.themoviedb.helper/resources/lib/api/fanarttv/api.py [hunk 2]
```diff
@@ -0,0 +1,117 @@
+from resources.lib.addon.plugin import get_language, get_setting
```
Comment (enen92) [inline plugin.video.themoviedb.helper/resources/lib/api/fanarttv/api.py:1]: the dummy init py file is no longer needed in python 3

File: plugin.video.themoviedb.helper/resources/lib/api/kodi/mapping.py [hunk 3]
```diff
@@ -37,27 +37,27 @@ def __init__(self, key=None):
                 'subkeys': ['metacritic', 'rating'],
```
Comment (enen92) [inline plugin.video.themoviedb.helper/resources/lib/api/kodi/mapping.py:37]: dummy init.py file is no longer needed in python3

File: plugin.video.themoviedb.helper/resources/lib/api/kodi/rpc.py [hunk 4]
```diff
@@ -176,15 +199,22 @@ def _get_database(self, dbtype, tvshowid=None, attempt_reconnect=False, logging=
         return self.get_database(dbtype, tvshowid, attempt_reconnect)
 
     def get_database(self, dbtype, tvshowid=None, attempt_reconnect=False, logging=True):
+        from xbmcvfs import Stat
+        cache_name = f'db.{dbtype}.{tvshowid}'
+        cache_data = self._cache.get_cache(cache_name)
+        db_updated = Stat('special://database/MyVideos119.db').st_mtime() or -1
```
Comment (enen92) [inline plugin.video.themoviedb.helper/resources/lib/api/kodi/rpc.py]: Looks like an hack to me. Is it really needed to stat our db files?
This is usually not allowed...

File: plugin.video.themoviedb.helper/resources/lib/api/omdb/api.py [hunk 5]
```diff
@@ -1,15 +1,19 @@
+from resources.lib.addon.plugin import get_setting
```
Comment (enen92) [inline plugin.video.themoviedb.helper/resources/lib/api/omdb/api.py:1]: please remove the dummy init.py file, no longer needed in python3

File: plugin.video.themoviedb.helper/resources/lib/api/tmdb/api.py [hunk 6]
```diff
@@ -1,40 +1,47 @@
-import xbmcgui
```
Comment (enen92) [inline plugin.video.themoviedb.helper/resources/lib/api/tmdb/api.py:1]: another dummy init py file

File: plugin.video.themoviedb.helper/resources/lib/api/trakt/api.py [hunk 7]
```diff
@@ -1,18 +1,20 @@
-import xbmc
```
Comment (enen92) [inline plugin.video.themoviedb.helper/resources/lib/api/trakt/api.py:1]: dummy init py above

File: plugin.video.themoviedb.helper/resources/lib/items/artselect.py [hunk 8]
```diff
@@ -0,0 +1,129 @@
+from xbmcgui import Dialog
```
Comment (enen92) [inline plugin.video.themoviedb.helper/resources/lib/items/artselect.py:1]: dummy init.py file above

File: plugin.video.themoviedb.helper/resources/lib/update/cacher.py [hunk 9]
```diff
@@ -1,5 +1,5 @@
-from resources.lib.files.utils import get_pickle, set_pickle
-from resources.lib.addon.timedate import is_future_timestamp, get_todays_date
+from resources.lib.files.futils import get_json_filecache, set_json_filecache
```
Comment (enen92) [inline plugin.video.themoviedb.helper/resources/lib/update/cacher.py:1]: dummy init.py file above

**Review by enen92 — COMMENTED** [review]
see comments

File: plugin.video.themoviedb.helper/resources/lib/api/kodi/rpc.py (hunk 4 above)
Comment (jurialmunkey) [reply plugin.video.themoviedb.helper/resources/lib/api/kodi/rpc.py]: It is definitely a hack. I build a dictionary based upon the database and then cache it for quick referencing against item IDs, so I only want to rebuild that dictionary if the library database changes. I couldn't find a way to get a "library last updated" type of value, hence the Stat. 

Happy to replace it with something else if there's a better option for checking when the library was updated last.

Comment (jurialmunkey) [reply plugin.video.themoviedb.helper/resources/lib/api/kodi/rpc.py]: Decided to remove the Stat and separate caching entirely. It only really has a speed advantage when the user has many thousands of items in their library, at which point their usage pattern is probably pretty questionable anyway...

File: plugin.video.themoviedb.helper/addon.xml (hunk 1 above)
Comment (jurialmunkey) [reply plugin.video.themoviedb.helper/addon.xml:50]: Left in mostly as a reminder that I do want to use this setting once some bugs that I've reported with this setting are fixed.

**Comment by jurialmunkey** [comment]
@enen92 - Fixed up those issues. Let me know if there's anything else needed.

**Review by pkscout — APPROVED** [review]

**Pull request merged**